doc/SAI-Proposal-HW-Resource.md

```diff
@@ -0,0 +1,155 @@
+
+PR #1482 introduces a method to
+1. Specify the stage of a resource
+2. Query an object's attribute stage for a given HW using capability API
+3. Error handling for incorrect resource stage specification
+
+## Common Data Type
+-----------------
+New data type sai_object_stage_t is introduced. This data type specifies if a resource is of type
+
+- SAI_OBJECT_STAGE_BOTH:
+  Resource is a common shared resource between ingress and egress HW pipeline
+- SAI_OBJECT_STAGE_INGRESS:
+  Resource belongs to ingress stage of HW pipeline
+- SAI_OBJECT_STAGE_EGRESS:
+  Resource belongs to egress stage of HW pipeline
+
+## Capability Query
+----------------
+Capability query API is enhanced with a query for attribute stage.NOS can query the attribute for a given HW support
+- SAI_ATTR_STAGE_BOTH: Attribute is common to ingress and egress stage
+- SAI_ATTR_STAGE_INGRESS: Attribute is applicable only to ingress stage
+- SAI_ATTR_STAGE_EGRESS: Attribute is applicabe only to egress stage
+
+## HOSTIF Trap Group
+-----------------
+A new CREATE_ONLY trap group attribute is introduced. NOS can set this attribute to specify the stage of the trap group.
+- SAI_HOTIF_TRAP_GROUP_ATTR_OBJECT_STAGE: 
+CREATE_ONLY attribute to configure a trap group of type stage ingress/egress/both
+
+## Policer Pool
+------------
+A new CREATE_ONLY policer attribute is introduced. NOS can set this attribute
+to specify the stage of the pool.
+- SAI_POLICER_ATTR_STAGE:
+  CREATE_ONLY attribute that specifies the stage of the policer pool. Default is kept as SAI_STAGE_BOTH common pool for backward compatibility.
+
+## Example
+-------
+Hostif trap's are implemented as an ingress or egress pipeline stage based on a HW architecture. In this example workflow, a hostif trap capability is queried. Based on returned value, appropriate trap group and policer pool is assigned to it. For e.g. a trap of type ingress will be assigned a trap group of type ingress using an ingress policer pool.
+
+## Sample Existing Workflow
+------------------------
+There is no concept of stages in this workflow.
+By default this trap is created in SAI_SWITCH_ATTR_DEFAULT_TRAP_GROUP trap group.
+
+```
+1. create_hostif_trap_group()
+2. create_hostif_trap(): 
+    By default this trap is created in 
+    SAI_SWITCH_ATTR_DEFAULT_TRAP_GROUP trap group.
+or
+2. create_hostif_trap():
+    For trap group created in step 1.
+3. create_policer()
+4. Attach policer to the trap using set of 
+    SAI_HOSTIF_TRAP_GROUP_ATTR_POLICER attribute.
+```
+
+### Sample New Workflow
+-------------------
+New workflow will identify if a given trap is applicable to ingress/egress/both. This will be done based on the enhanced capability query API.
+```
+1. Query the HW capability for a given trap
+    sai_status_t sai_query_attribute_capability(
+        switch_id,
+        SAI_OBJECT_TYPE_HOSTIF_TRAP,
+        SAI_HOSTIF_TRAP_TYPE_DNAT_MISS,

```

Comment:
As discussed in the community meeting, this would not work since the API takes as input a sai_attr_id_t. We will not be able to pass the actual trap_type.

One solution, would be to create a new API that takes as input a sai_attribute_t and returns the stage. 

```
sai_status_t sai_query_object_stage(
        _In_ sai_object_id_t switch_id,
        _In_ sai_object_type_t object_type,
        _In_ uint32_t attr_count,
        _In_ const sai_attribute_t *attr_list,
        _Out_ sai_object_stage_t *stage);
```

Reply:
Ravi, Thanks for the feedback.
Incorporated the suggested change.
https://github.com/opencomputeproject/SAI/pull/1528/commits/bbb458586b22832563ca663a3695fb5df46f1660

Reply:
Thanks. Please update this section of the document.